mdd/python/translation/openconfig_xe/xe_qos.py

```diff
@@ -0,0 +1,280 @@
+# -*- mode: python; python-indent: 4 -*-
+from translation.openconfig_xe.common import xe_system_get_interface_ip_address
+from translation.common import get_interface_type_and_number
+
+dscp_dict = {8:'cs1', 10:'af11', 12:'af12', 14:'af13', 16:'cs2', 18:'af21', 20:'af22',
+                22:'af23', 24:'cs3', 26:'af31', 28:'af32', 30:'af33', 32:'cs4', 34:'af41',
+                36:'af42', 38:'af43', 40:'cs5', 46:'ef', 48:'cs6', 56:'cs7', 0:'default'}
+
+def xe_qos_program_service(self, nso_props) -> None:
+    """
+    Program service
+    """
+
+    device_cdb = nso_props.root.devices.device[nso_props.device_name].config
+
+    # Forwarding-groups
+    """
+    Configure forwarding-groups
+    """
+    for fg in nso_props.service.oc_qos__qos.forwarding_groups.forwarding_group:
+        if device_cdb.ios__policy_map.exists(fg.name):
+            del device_cdb.ios__policy_map[fg.name]
+        device_cdb.ios__policy_map.create(fg.name)
+
+    # Class-map
+    """
+    Configure classifiers
+    """
+    list_cmap = []
+    for c_map in nso_props.service.oc_qos__qos.classifiers.classifier:
+        pmap_cmap = {}
+        # Configure class-map
+        if device_cdb.ios__class_map.exists(c_map.name) and c_map.name != 'class-default':
+            del device_cdb.ios__class_map[c_map.name]
+        elif c_map.name == 'class-default':
+            for t_map in c_map.terms.term:
+                pmap_cmap[t_map.actions.config.target_group] = c_map.name
+        else:
+            device_cdb.ios__class_map.create(c_map.name)
+            for t_map in c_map.terms.term:
+                pmap_cmap[t_map.actions.config.target_group] = c_map.name
+                # Configure "match ip dscp"
```

Comment:
"ip dscp" or "dscp" (for IPv4 and IPv6) should be based on the c_map.type. The c_map.type will determine the availability of "t_map.conditions.[ ipv4, ipv6, etc...]"

Reply:
Do you mean:
if type: ETHERNET ---> use 'dscp'
if type: IPv4 ---> use 'ip dscp'
or something else?

Reply:
In this code:
    for c_map in nso_props.service.oc_qos__qos.classifiers.classifier:
        pmap_cmap = {}
        # Configure class-map
        if device_cdb.ios__class_map.exists(c_map.name) and c_map.name != 'class-default':
            del device_cdb.ios__class_map[c_map.name]
        elif c_map.name == 'class-default':
            for t_map in c_map.terms.term:
                pmap_cmap[t_map.actions.config.target_group] = c_map.name
        else:
            device_cdb.ios__class_map.create(c_map.name)
            for t_map in c_map.terms.term:
                pmap_cmap[t_map.actions.config.target_group] = c_map.name
                # Configure "match ip dscp"
                if t_map.conditions.ipv4.config.protocol == 4:
                    # Configure multiple dscp statements
                    if t_map.conditions.ipv4.config.dscp_set:

The code is already looking into IPv4 terms (if t_map.conditions.ipv4.config.protocol == 4:) to determine if the term is using IPv4, but the term could be IPv6. For instance if c_map.type == "IPv6" (or something like that), then the term would be t_map.conditions.ipv6....